app/models/route.rb

```diff
@@ -84,62 +83,66 @@ def default_segments_mode
   private
 
   def validate_incoming_path
-    unless valid_local_path?(self.incoming_path)
-      errors[:incoming_path] << "is not a valid absolute URL path"
-    end
+    errors[:incoming_path] << "must start with /" unless
+      incoming_path.starts_with?("/")
+
+    uri = URI.parse(incoming_path)
+
+    errors[:incoming_path] << "cannot end with /" if
+      uri.path != "/" && uri.path.end_with?("/")
+    errors[:incoming_path] << "cannot contain //" if uri.path =~ %r{//}
+
+    errors[:incoming_path] << "does not equal the URI path" unless
+      uri.path == incoming_path
+  rescue URI::InvalidURIError
+    errors[:incoming_path] << "is an invalid URI"
   end
 
   def validate_redirect_to
     return unless self.redirect_to.present? # This is to short circuit nil values
-    if self.segments_mode == 'ignore'
-      unless valid_ignore_redirect_target?(self.redirect_to)
-        errors[:redirect_to] << "is not a valid redirect target"
-      end
+
+    if self.redirect_to.starts_with?("/")
+      validate_internal_target?(self.redirect_to)
     else
-      unless allow_segments?(self.redirect_to)
-        errors[:redirect_to] << "is not a valid redirect target"
-      end
+      validate_external_target?(self.redirect_to)
     end
-  end
 
-  def allow_segments?(url)
-    valid_local_path?(url) || valid_whitelisted_url?(url)
-  end
+    if segments_mode == "preserve"
+      uri = URI.parse(redirect_to)
+
+      if uri.fragment.present?
+        errors[:redirect_to] << "cannot contain fragment if the segments mode is preserve"
+      end
 
-  def valid_whitelisted_url?(url)
-    uri = URI.parse(url)
-    return false unless uri.absolute? && path_is_root_or_empty?(uri)
-    belongs_to_whitelist?(uri)
+      if uri.query.present?
+        errors[:redirect_to] << "cannot contain query parameters if the segments mode is preserve"
+      end
+    end
   rescue URI::InvalidURIError
-    false
+    errors[:redirect_to] << "is an invalid URI"
   end
 
-  def path_is_root_or_empty?(uri)
-    uri.path.blank? || uri.path == '/'
-  end
+  def validate_external_target?(target)
```

Comment:
This method and the one following shouldn't end with `?`; they don't return whether the target is valid (in which case they would be called `valid_internal_target?`, they actually do the validation.